apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -101,61 +112,71 @@
-foreground_output['pythonNames'] = foreground_output['output']
-
-# Modify output to reflect blockly id names.
-background_output['output'] = [background_effects_map[bg] for bg in background_output['output']]
-foreground_output['output'] = [foreground_effects_map[fg] for fg in foreground_output['output']]
-palette_output['output'] = [palettes_map[pal] for pal in palette_output['output']]
-
-with open("apps/static/dance/ai/model/cached-spacy-palette-map.json", "w") as json_file:
-    json_file.write(json.dumps(palette_output))
-
-with open("apps/static/dance/ai/model/cached-spacy-background-map.json", "w") as json_file:
-    json_file.write(json.dumps(background_output))
-
-with open("apps/static/dance/ai/model/cached-spacy-foreground-map.json", "w") as json_file:
-    json_file.write(json.dumps(foreground_output))
+
+# Create maps of model_descriptive_name to blockly_id and print lists
+# as 'model_descriptive_name | blockly_id | blockly_user_facing_name'
+# Updates lists in comments at bottom of this file.
+create_map_print_options(PALETTES, model_descriptive_names_map[PALETTES], palette_blockly_ids_list, palette_user_facing_names_list),
+create_map_print_options(BACKGROUND_EFFECTS, model_descriptive_names_map[BACKGROUND_EFFECTS], background_effect_blockly_ids_list, background_effect_user_facing_names_list),
+create_map_print_options(FOREGROUND_EFFECTS, model_descriptive_names_map[FOREGROUND_EFFECTS], foreground_effect_blockly_ids_list, foreground_effect_user_facing_names_list)
+
+# We use cache files to store ada's raw embedding outputs to reduce Open AI query costs.
+# An embedding is a vector (list) of floating point numbers, and embeddings are used to
+# measure the relatedness of text strings. The distance between two vectors measures
+# their relatedness. https://platform.openai.com/docs/guides/embeddings
+# These caches are stored as pickle files; python's native way to serialize data.
+embeddings_paths = {
+    EMOJIS: 'apps/static/dance/ai/model/input_embeddings.pkl',
+    PALETTES: 'apps/static/dance/ai/model/palette_embeddings.pkl',
+    BACKGROUND_EFFECTS: 'apps/static/dance/ai/model/background_embeddings.pkl',
+    FOREGROUND_EFFECTS: 'apps/static/dance/ai/model/foreground_embeddings.pkl'
+}
+
+embeddings_caches = {}
+for path in embeddings_paths.keys():
+    embeddings_caches[path] = load_embeddings_cache(path)
+
+# Retrieve embeddings using model_descriptive_names for input emojis, palettes, background effects, and foreground effects.
+options_lists = {
+    EMOJIS: emoji_model_descriptive_names_list,
+    PALETTES: model_descriptive_names_map[PALETTES],
+    BACKGROUND_EFFECTS: model_descriptive_names_map[BACKGROUND_EFFECTS],
+    FOREGROUND_EFFECTS: model_descriptive_names_map[FOREGROUND_EFFECTS]
+}
+
+embeddings = {}
+# embeddings is a dictionary with four lists of embeddings:
+# EMOJIS: [[emoji1_embedding], [emoji2_embedding], ...]
+# PALETTES: [[palette1_embedding], [palette2_embedding], ...]
```

Comment:
Nit: I think there's maybe a little bit of a "less is more" thing with some of the comments...I think I could get the idea with a less verbose comment here (and maybe a couple other places as well)?

Reply:
Sure - that makes sense. Would it be useful to be share this level of details in PR description?

Reply:
Yeah maybe, I was thinking the PR description could use a bit more detail. I also had the thought reading through this that a lot of the comments you've added might add up nicely to a README that gives an overview of the process here. Whatever you think is best!

Reply:
I really like the idea of a README for running of the script. Great idea!
@tshaffercodeorg is going to update the PR description and add testing results and why we're moving to Ada. Thanks so much!

Reply:
Definitely big +1 on improving the documentation, I've included a more detailed write-up in the PR summary to help clarify the decisions and provide just a cleaner overview of the work being done!

Reply:
PR description is super helpful, thanks @tshaffercodeorg !